### PR TITLE
SOLR-15843 Update Log4J to 2.15

### PR DESCRIPTION
**Pull request by madrob**
https://issues.apache.org/jira/browse/SOLR-15843

### CODE REVIEW

**Comment by ramenjosh** [comment]
Hey team, can this also be backported to Solr 7.7 as a new 7.7.4 release? We are currently running 7.7.3 and need this urgently.

**Comment by ramenjosh** [comment]
Hey team, can this also be backported to 7.7 as a new 7.7.4 release? Many people still running Solr 7 will also require this fix.

**Review by janhoy — APPROVED** [review]
Looks good, you can consider moving the changes entry to bugfix as our reason to upgrade is not to use new features but to get bugfixes...

**Comment by uschindler** [comment]
> Hey team, can this also be backported to 7.7 as a new 7.7.4 release? Many people still running Solr 7 will also require this fix.

This won't happen anymore as 7.x is out of maintenance. To fix it, just modify your local solr.in.sh and pass the extra command line parameter: `-Dlog4j2.formatMsgNoLookups=true` 

We will add this to the release notes.

**Review by uschindler — APPROVED** [review]
Please add the above instructions to release notes, so we inform users how to fix older versions (like 7.x, 6.x,...).

**Comment by madrob** [comment]
Release notes or the security page on the site (or both?)

Solr 7 is affected but AFAICT Solr 5 and 6 are not because they use log4j 1.2.17

**Comment by uschindler** [comment]
> Release notes or the security page on the site (or both?)
> 
> Solr 7 is affected but AFAICT Solr 5 and 6 are not because they use log4j 1.2.17

I would maybe do both. Let's add a notice with the fix to our news list (under security) and add the release note later when we publish a release.

In addition, depending on your configuration (not the default), log4j v1 is also affected. Theres a special appender doing the same style of JNDI stuff: https://github.com/apache/logging-log4j2/pull/608#issuecomment-990494126

**Comment by plumdog** [comment]
Should we expect patched releases for older 8.x versions? In my case, 8.9? Or will I have to take 8.11?

**Comment by plumdog** [comment]
Also, regarding Solr 5 and 6 and log4j: https://github.com/apache/logging-log4j2/pull/608#issuecomment-990494126

Based on that, I think they would be affected.

**Comment by madrob** [comment]
> Should we expect patched releases for older 8.x versions? In my case, 8.9? Or will I have to take 8.11?

You should not expect additional 8.9.z releases

**Comment by uschindler** [comment]
> Also, regarding Solr 5 and 6 and log4j: apache/logging-log4j2#608 (comment)

This is only a problem if you use special appenders, which solr does not do.

**Comment by acdha** [comment]
I was wondering whether someone could confirm that the top-level version lock will also affect the contrib prometheus-exporter. My scan of our container images turned up a second copy of log4j 2.14.1 in `/opt/solr-8.11.0/contrib/prometheus-exporter/lib/`.

**Comment by plumdog** [comment]
> You should not expect additional 8.9.z releases

@madrob So if I want something both stable and patched, I'll need 8.11.next?

**Comment by uschindler** [comment]
> > You should not expect additional 8.9.z releases
> 
> @madrob So if I want something both stable and patched, I'll need 8.11.next?

I would NOT wait for a release or asking the same question like this one all the time (ask them one time on mailing list please) and instead act NOW:

Just open your `solr.in.sh` in your facorite editor and add `SOLR_OPTS="$SOLR_OPTS -Dlog4j2.formatMsgNoLookups=true"`. Restart Solr. Done, all safe.

File: solr/bin/solr.in.cmd [hunk 1]
```diff
@@ -213,8 +213,8 @@ REM set SOLR_ADMIN_UI_DISABLED=false
 REM Solr is by default allowed to read and write data from/to SOLR_HOME and a few other well defined locations
 REM Sometimes it may be necessary to place a core or a backup on a different location or a different disk
 REM This parameter lets you specify file system path(s) to explicitly allow. The special value of '*' will allow any path
-REM SOLR_OPTS="%SOLR_OPTS% -Dsolr.allowPaths=D:\,E:\other\path"
+REM SOLR_OPTS=%SOLR_OPTS% -Dsolr.allowPaths=D:\,E:\other\path
```
Comment (uschindler) [inline solr/bin/solr.in.cmd]: there must be a "set" before the SOLR_OPTS (in both cases)!

**Review by uschindler — CHANGES_REQUESTED** [review]

**Comment by uschindler** [comment]
> > > You should not expect additional 8.9.z releases
> > 
> > 
> > @madrob So if I want something both stable and patched, I'll need 8.11.next?
> 
> I would NOT wait for a release or asking the same question like this one all the time (ask them one time on mailing list please) and instead act NOW:
> 
> Just open your `solr.in.sh` in your favorite editor and add `SOLR_OPTS="$SOLR_OPTS -Dlog4j2.formatMsgNoLookups=true"`. Restart Solr. Done, all safe.

Here is the security advisory, it shows you how to harden your solr server: https://solr.apache.org/security.html#apache-solr-affected-by-apache-log4j-cve-2021-44228-jndi-features-do-not-protect-against-attacker-controlled-ldap-and-other-jndi-related-endpoints

**Review by uschindler — APPROVED** [review]

**Comment by uschindler** [comment]
Hi,

Solr does use MDC (the %X pattern), but the values are not user generated and all come from config files and are enforced to comply to certain formats (e.g., no $ possible). Shard, replica, collection names are sanitized.

In short all fine, no need to change the mitigation instructions. There is also no need to update log4j in older versions of Solr.

Uwe

**Comment by mario-canva** [comment]
Thanks @uschindler appreciate the quick response! However, their advisory also states other attack vectors may be possible:

**Comment by uschindler** [comment]
The other attack vectors are also not possible with Solr:

- Logger.printf("%s", userInput) is not used
- custom message factory is not used

Uwe

**Comment by mario-canva** [comment]
Understood. I still think pays off being a bit conservative here, it is a critical CVE after all and the log4j team stated other attack vectors may be possible (beyond the ones they have found already).

Internally we will follow Log4j advice to not use that flag as mitigation and remove the `JndiLookup` class from the jar.

Thanks!

**Pull request merged**